inspirehep/modules/arxiv/ext.py

```diff
@@ -0,0 +1,168 @@
+        self.app = app
+        if app is not None:
+            self.init_app(app)
+
+    def init_app(self, app):
+        """Initialize a Flask application."""
+        app.config.setdefault("ARXIV_API_URL",
+                              "http://export.arxiv.org/oai2")
+        app.config.setdefault("ARXIV_ENDPOINT", "_arxiv.search")
+        app.config.setdefault("ARXIV_URL_RULE", "/arxiv/search")
+
+        # Follow the Flask guidelines on usage of app.extensions
+        if not hasattr(app, "extensions"):
+            app.extensions = {}
+        if "arxiv" in app.extensions:
+            raise RuntimeError("Flask application already initialized")
+        app.extensions["arxiv"] = self
+
+        if app.config["ARXIV_ENDPOINT"]:
+            app.add_url_rule(app.config["ARXIV_URL_RULE"],
+                             app.config["ARXIV_ENDPOINT"],
+                             login_required(self.search))
+
+    def get_response(self, arxiv_id):
+        """Get ArXiv response from the ``ARXIV_API_URL`` page."""
+        response = requests.get(
+            current_app.config["ARXIV_API_URL"],
+            params=dict(
+                verb="GetRecord",
+                metadataPrefix="arXiv",
+                identifier="oai:arXiv.org:{term}".format(
+                    term=arxiv_id.strip()),
+            ),
+        )
+        return response
+
+    def get_json(self, arxiv_id):
+        """Get ArXiv json data."""
+        response = self.get_response(arxiv_id)
+        data = etree_to_dict(fromstring(response.content))
+        query = {}
+
+        for d in data["OAI-PMH"]:
+            query.update(dict(d.items()))
+        del data["OAI-PMH"]
+
+        if "error" in query:
+            # Check if the ArXiv ID is malformed:
+            if query["error"].startswith("Malformed"):
+                query = {}
+                data["status"] = "malformed"
+            # Check if the ArXiv ID has version specified:
+            elif 'versions' in query["error"]:
+                query = {}
+                data["status"] = "unsupported_versioning"
+            # Otherwise, ArXiv ID was not found:
+            else:
+                query = {}
+                data["status"] = "notfound"
+        else:
+            for d in query['GetRecord'][0][
+                    'record'][1]['metadata'][0]['arXiv']:
+                query.update(dict(d.items()))
+            del query['GetRecord']
+            data["status"] = "success"
+
+        data["source"] = "arxiv"
+        data["query"] = query
+
+        return data
+
+    def search(self, arxiv=None):
+        """Search for given ArXiv ID."""
+        arxiv = arxiv or request.args.get("arxiv")
+
+        # query arxiv
+        result = self.get_json(arxiv)
+
+        resp = jsonify(result)
+        resp.status_code = response_code.get(result['status'],
+                                             result['status'])
+        return resp
+
+__all__ = ("Arxiv", )
```

Comment:
I think this whole part should be actually factor out in an dedicated repo with nice tests.